### PR TITLE
Refactor: make putTerm in Transaction, not IO

### PR DESCRIPTION
**Pull request by mitchellwrosen**
- [x] Depends on #3569 

---

## Overview

This PR makes `Codebase.putTerm` in `Transaction`, not `IO`

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase.hs [hunk 1]
```diff
@@ -290,15 +290,15 @@ installUcmDependencies c = do
 -- if it makes sense to later.
 addDefsToCodebase ::
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase.hs:291]: Is the goal for this to also be in Transaction then?
I'm  guessing we just need to also convert some more methods before we can do that?

**Review by ChrisPenner — APPROVED** [review]

File: parser-typechecker/src/Unison/Codebase.hs (hunk 1 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Codebase.hs:291]: Yep.

File: unison-cli/src/Unison/Codebase/Editor/Propagate.hs [hunk 2]
```diff
@@ -393,20 +393,17 @@ propagate patch b = case validatePatch patch of
...
-                          <$> componentMap
-                      seen' = seen <> Set.fromList (view _1 <$> Map.elems componentMap)
-                  mayComponent <- verifyTermComponent codebase componentMap' es
-                  pure (componentMap, seen', mayComponent)
+              componentMap <- unhashTermComponent codebase r
+              let componentMap' =
+                    over
+                      _2
+                      (Term.updateDependencies termReplacements typeReplacements)
+                      <$> componentMap
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/Propagate.hs]: Do you want this here, or local to the definition of `mayComponent`?

**Review by aryairani — APPROVED** [review]

File: unison-cli/src/Unison/Codebase/Editor/Propagate.hs (hunk 2 above)
Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/Propagate.hs]: I think its scope could be pushed down into a let, but I'm not sure it matters much. It was originally just inline like this (before a recent PR that indented due to a `runTransaction` call, and I can see that the corresponding `doType` above also just defines it inline.

Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/Propagate.hs]: Fixed

**Pull request merged**